internal_ws/ykcompile/src/arch/x86_64/mod.rs

```diff
@@ -123,45 +123,121 @@ macro_rules! binop_add_sub {
+                    // Handle cases where the second operand was stored in an already clobbered
+                    // register. In these cases we load the value back from the stack.
+                    let (r, borrow_rcx) = if src_r == RAX.code() {
+                        let rax_off = i32::try_from(QWORD_REG_SIZE).unwrap();
                         dynasm!(self.asm
-                            ; $op Rb(r)
+                            ; push rcx
+                            ; mov rcx, [rsp + rax_off]
                         );
-                    }
-                    2 => {
+                        (RCX.code(), true)
+                    } else if size > 1 && src_r == RDX.code() {
+                        let rdx_off = i32::try_from(QWORD_REG_SIZE).unwrap() * 2;
                         dynasm!(self.asm
-                            ; $op Rw(r)
+                            ; push rcx
+                            ; mov rcx, [rsp + rdx_off]
                         );
-                    }
-                    4 => {
+                        (RCX.code(), true)
+                    } else {
+                        (src_r, false)
+                    };
+                    // Zero unused extended operand space.
+                    if size > 1 {
                         dynasm!(self.asm
-                            ; $op Rd(r)
+                            ; xor rdx, rdx
                         );
                     }
-                    8 => {
+                    match size {
+                        1 => {
+                            dynasm!(self.asm
+                                ; and ax, 0xff // Zero unused extended operand space.
+                                ; $op Rb(r)
+                            );
+                        }
+                        2 => {
+                            dynasm!(self.asm
+                                ; $op Rw(r)
+                            );
+                        }
+                        4 => {
+                            dynasm!(self.asm
+                                ; $op Rd(r)
+                            );
+                        }
+                        8 => {
+                            dynasm!(self.asm
+                                ; $op Rq(r)
+                            );
+                        }
+                        _ => unreachable!(format!("{}", SIR.ty(&opnd2.ty()))),
+                    }
+                    // If we borrowed RCX, then restore it.
+                    if borrow_rcx {
                         dynasm!(self.asm
-                            ; $op Rq(r)
+                            ; pop rcx
                         );
                     }
-                    _ => unreachable!(format!("{}", SIR.ty(&opnd2.ty()))),
                 },
                 Location::Mem(..) => todo!(),
                 Location::Const { val, .. } => {
+                    dynasm!(self.asm // FIXME not always.
+                        ; xor rdx, rdx
```

Comment:
This doesn't look right. If the operand is a constant, we zero `rdx`, but we only `push/pop rdx` if the operand size is 1. What happens with constants of bigger sizes? Do we then clobber `rdx` without restoring it?

Reply:
You're right, it should be guarded by `if size >= 1`.

I've not yet fixed this because I'd like to find a test which fails under the current code.

Reply:
`>=` or just `>`?

Reply:
yes, `>`.

I'm struggling to get a test that hits this bug. The problem is that our unoptimised MIR copies a lot, and it's always an old unused copy that gets corrupted...

I think we should move on, but we could fix the bug and import the passing test anyway?

Reply:
Well 7e87b29 fixes the bug and adds a test which **might** have triggered the bug, had our MIR and SIR been more optimal.